paddle/fluid/operators/elementwise/elementwise_op.h

```diff
@@ -250,34 +252,23 @@ class ElemwiseGradKernel : public framework::OpKernel<T> {
   }
 };
 
-class ElementwiseOpInplace : public framework::InplaceInToOut {
+class ElementwiseOpInplace : public framework::InplaceOpInference {
  public:
-  using framework::InplaceInToOut::InplaceInToOut;
-
- protected:
-  std::unordered_map<std::string, std::string> Apply(
-      const framework::OpDesc &op_desc,
-      framework::BlockDesc *block) const override {
+  std::unordered_map<std::string, std::string> operator()(
+      const framework::OpDesc &op_desc) const override {
     return std::unordered_map<std::string, std::string>{
         {"X", "Out"},
     };
   }
 };
 
-class ElementwiseGradOpInplace : public framework::InplaceInToOut {
+class ElementwiseGradOpInplace : public framework::InplaceOpInference {
  public:
-  using framework::InplaceInToOut::InplaceInToOut;
-
- protected:
-  std::unordered_map<std::string, std::string> Apply(
-      const framework::OpDesc &op_desc,
-      framework::BlockDesc *block) const override {
-    std::unordered_map<std::string, std::string> ret;
-    if (block->HasVar(framework::GradVarName("X")) &&
-        block->HasVar(framework::GradVarName("Out"))) {
-      ret[framework::GradVarName("Out")] = framework::GradVarName("X");
-    }
-    return ret;
+  std::unordered_map<std::string, std::string> operator()(
+      const framework::OpDesc &op_desc) const override {
+    return std::unordered_map<std::string, std::string>{
+        {framework::GradVarName("Out"), framework::GradVarName("X")},
```

Comment:
Is it safe to remove the block hasvar check?

Reply:
> Is it safe to remove the block hasvar check?

block_desc is used here to get the Variable name here, but actually we can get the variable name in Node structure. So now I have changed logic like this:
the operator configure the potential parameters which can be apply the in-place strategy like:
X => Out
the inplace_op_pass which can access the real node can get the variable by parameter name can do the runtime analysis to determine we apply the in-place or not.
So this part of logic has been moved to inplace_op_pass

Reply:
related icafe ticket: http://newicafe.baidu.com:80/issue/DLTP-667/show?from=page